File: packages/py/glin_profanity/utils/variant_mapping.py
"""Map span positions from normalized variant text back to the original."""

from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass

from glin_profanity.utils.leetspeak import AGGRESSIVE_SUBSTITUTIONS, MODERATE_SUBSTITUTIONS
from glin_profanity.utils.unicode import homoglyph_to_ascii

_LEET_TO_ASCII = {
    "@": "a",
    "$": "s",
    "!": "i",
    "1": "i",
    "0": "o",
    "3": "e",
    "4": "a",
    "5": "s",
    "7": "t",
    "8": "b",
    "9": "g",
}

_LEET_SUBSTITUTIONS: dict[str, str] = {
    **MODERATE_SUBSTITUTIONS,
    **AGGRESSIVE_SUBSTITUTIONS,
}

_SKIPPABLE_ORIGINAL_CHARS = {"*", ".", "_", "-", " "}

# Keep leetspeak/masking symbols when they belong to the obfuscated token.
_EDGE_MASKING_CHARS = frozenset("@$!#*")


@dataclass(frozen=True)
class OriginalSpan:
    start: int
    end: int
    matched_text: str


def _normalize_char_for_align(char: str) -> str:
    mapped = homoglyph_to_ascii(char)
    decomposed = unicodedata.normalize("NFKD", mapped)
    base = "".join(c for c in decomposed if unicodedata.combining(c) == 0)
    return base.lower()


def _chars_equal(left: str, right: str) -> bool:
    return left == right or left.lower() == right.lower()


def _chars_align(original_char: str, variant_char: str) -> bool:
    if _chars_equal(original_char, variant_char):
        return True

    original_norm = _normalize_char_for_align(original_char)
    variant_norm = _normalize_char_for_align(variant_char)
    if original_norm and original_norm == variant_norm:
        return True

    mapped = _LEET_TO_ASCII.get(original_char) or _LEET_TO_ASCII.get(
        original_char.lower()
    )
    if mapped is not None and mapped.lower() == variant_norm:
        return True

    substituted = _LEET_SUBSTITUTIONS.get(original_char) or _LEET_SUBSTITUTIONS.get(
        original_char.lower()
    )
    if substituted is not None and substituted.lower() == variant_norm:
        return True

    homoglyph = homoglyph_to_ascii(original_char)
    if homoglyph != original_char and _normalize_char_for_align(homoglyph) == variant_norm:
        return True

    return False


def _is_skippable_original_char(char: str) -> bool:
    return char in _SKIPPABLE_ORIGINAL_CHARS


def _is_combining_mark(char: str) -> bool:
    return bool(char) and unicodedata.combining(char) != 0


def _should_trim_edge_punctuation(char: str) -> bool:
    if char in _EDGE_MASKING_CHARS:
        return False
    if char in {".", "_", "-"}:
        return True
    category = unicodedata.category(char)
    return category.startswith("P") or category.startswith("Z")


def trim_profane_span_edges(text: str, start: int, end: int) -> OriginalSpan:
    """Drop leading/trailing whitespace and outer punctuation from a profane span."""
    if start >= end:
        return OriginalSpan(start=start, end=end, matched_text="")

    while start < end and text[start].isspace():
        start += 1
    while start < end and text[end - 1].isspace():
        end -= 1

    while start < end and text[start] in {".", "_", "-"}:
        start += 1
    while start < end and text[end - 1] in {".", "_", "-"}:
        end -= 1

    while start < end and _should_trim_edge_punctuation(text[start]):
        start += 1
    while start < end and _should_trim_edge_punctuation(text[end - 1]):
        end -= 1

    return OriginalSpan(start=start, end=end, matched_text=text[start:end])


def _finalize_span(original: str, start: int, end: int) -> OriginalSpan:
    return trim_profane_span_edges(original, start, end)


def _fallback_span(
    original: str, variant: str, variant_start: int, variant_end: int
) -> OriginalSpan:
    needle = variant[variant_start:variant_end]
    if not needle:
        return OriginalSpan(start=0, end=0, matched_text="")

    idx = original.lower().find(needle.lower())
    if idx >= 0:
        return _finalize_span(original, idx, idx + len(needle))

    start = min(variant_start, len(original))
    end = min(variant_end, len(original))
    return _finalize_span(original, start, end)


def map_variant_span_to_original(
    original: str,
    variant: str,
    variant_start: int,
    variant_end: int,
) -> OriginalSpan:
    """Locate the original span for ``variant[variant_start:variant_end]``."""
    if variant_start >= variant_end:
        return OriginalSpan(start=0, end=0, matched_text="")

    if original == variant:
        return _finalize_span(original, variant_start, variant_end)

    original_index = 0
    variant_index = 0
    orig_start = -1
    orig_end = -1

    while variant_index < len(variant) and original_index <= len(original):
        if variant_index == variant_start and orig_start == -1:
            orig_start = original_index
        if variant_index == variant_end:
            orig_end = original_index
            break

        if original_index >= len(original):
            variant_index += 1
            continue

        variant_char = variant[variant_index]
        original_char = original[original_index]

        if _is_combining_mark(original_char):
            original_index += 1
            continue

        if _chars_align(original_char, variant_char):
            variant_index += 1
            original_index += 1
            continue

        if _is_skippable_original_char(original_char):
            original_index += 1
            continue

        original_index += 1

    if orig_end == -1 and variant_index >= variant_end:
        orig_end = original_index

    if orig_start == -1 or orig_end == -1:
        return _fallback_span(original, variant, variant_start, variant_end)

    if orig_start >= orig_end:
        return _fallback_span(original, variant, variant_start, variant_end)

    matched_text = original[orig_start:orig_end]
    if not matched_text:
        return _fallback_span(original, variant, variant_start, variant_end)

    return _finalize_span(original, orig_start, orig_end)


ProfaneSpan = tuple[str, int, int]


def dedupe_profane_spans_by_overlap(spans: list[ProfaneSpan]) -> list[ProfaneSpan]:
    """Keep longest original-text span when starts or ranges overlap."""
    if not spans:
        return []

    ordered = sorted(spans, key=lambda item: (item[1], -(item[2] - item[1])))
    kept: list[ProfaneSpan] = []

    for candidate in ordered:
        word, start, end = candidate
        length = end - start

        if any(
            start >= kept_start
            and end <= kept_end
            and (kept_end - kept_start) > length
            for _, kept_start, kept_end in kept
        ):
            continue

        kept = [
            item
            for item in kept
            if not (
                item[1] >= start
                and item[2] <= end
                and length > (item[2] - item[1])
            )
        ]
        kept.append((word, start, end))

    return kept


_NESTED_WORD_BOUNDARY = re.compile(r"\w")


def is_nested_profane_span(shorter: str, longer: str) -> bool:
    """True when shorter is a word-bounded substring of longer."""
    if len(longer) <= len(shorter):
        return False

    search_from = 0
    while search_from <= len(longer) - len(shorter):
        idx = longer.find(shorter, search_from)
        if idx == -1:
            return False

        before_ok = idx == 0 or not _NESTED_WORD_BOUNDARY.match(longer[idx - 1])
        after_idx = idx + len(shorter)
        after_ok = after_idx == len(longer) or not _NESTED_WORD_BOUNDARY.match(
            longer[after_idx]
        )

        if before_ok and after_ok:
            return True

        search_from = idx + 1

    return False

File: packages/js/src/utils/variantMapping.ts
/**
 * Maps span positions from a normalized variant back to the original text.
 * Assumes normalization is order-preserving (no reordering of characters).
 */

import { hasNonZeroCombiningClass } from './combiningClass';
import { AGGRESSIVE_SUBSTITUTIONS, MODERATE_SUBSTITUTIONS } from './leetspeak';
import { homoglyphToAscii } from './unicode';
import { isUnicodeWordChar } from './wordScript';

export interface OriginalSpan {
  start: number;
  end: number;
  matchedText: string;
}

const LEET_TO_ASCII: Record<string, string> = {
  '@': 'a',
  $: 's',
  '!': 'i',
  '1': 'i',
  '0': 'o',
  '3': 'e',
  '4': 'a',
  '5': 's',
  '7': 't',
  '8': 'b',
  '9': 'g',
};

const LEET_SUBSTITUTIONS: Record<string, string> = {
  ...MODERATE_SUBSTITUTIONS,
  ...AGGRESSIVE_SUBSTITUTIONS,
};

const SKIPPABLE_ORIGINAL_CHARS = new Set(['*', '.', '_', '-', ' ']);

/** Keep leetspeak/masking symbols when they belong to the obfuscated token. */
const EDGE_MASKING_CHARS = new Set(['@', '$', '!', '#', '*']);

function normalizeCharForAlign(char: string): string {
  const mapped = homoglyphToAscii(char);
  const decomposed = mapped.normalize('NFKD');
  let base = '';
  for (const codePoint of decomposed) {
    if (!hasNonZeroCombiningClass(codePoint)) {
      base += codePoint;
    }
  }
  return base.toLowerCase();
}

function charsEqual(a: string, b: string): boolean {
  return a === b || a.toLowerCase() === b.toLowerCase();
}

function charsAlign(originalChar: string, variantChar: string): boolean {
  if (charsEqual(originalChar, variantChar)) {
    return true;
  }

  const originalNorm = normalizeCharForAlign(originalChar);
  const variantNorm = normalizeCharForAlign(variantChar);
  if (originalNorm && originalNorm === variantNorm) {
    return true;
  }

  const leet = LEET_TO_ASCII[originalChar] ?? LEET_TO_ASCII[originalChar.toLowerCase()];
  if (leet !== undefined && leet.toLowerCase() === variantNorm) {
    return true;
  }

  const substituted =
    LEET_SUBSTITUTIONS[originalChar] ?? LEET_SUBSTITUTIONS[originalChar.toLowerCase()];
  if (substituted !== undefined && substituted.toLowerCase() === variantNorm) {
    return true;
  }

  const homoglyph = homoglyphToAscii(originalChar);
  if (homoglyph !== originalChar && normalizeCharForAlign(homoglyph) === variantNorm) {
    return true;
  }

  return false;
}

function isSkippableOriginalChar(char: string): boolean {
  return SKIPPABLE_ORIGINAL_CHARS.has(char);
}

function isCombiningMark(char: string): boolean {
  return hasNonZeroCombiningClass(char);
}

function shouldTrimEdgePunctuation(char: string): boolean {
  if (EDGE_MASKING_CHARS.has(char)) {
    return false;
  }
  if ('._-'.includes(char)) {
    return true;
  }
  return /\p{P}|\p{Z}/u.test(char);
}

/** Drop leading/trailing whitespace and outer punctuation from a profane span. */
export function trimProfaneSpanEdges(text: string, start: number, end: number): OriginalSpan {
  if (start >= end) {
    return { start, end, matchedText: '' };
  }

  while (start < end && /\s/u.test(text[start]!)) {
    start++;
  }
  while (start < end && /\s/u.test(text[end - 1]!)) {
    end--;
  }

  while (start < end && '._-'.includes(text[start]!)) {
    start++;
  }
  while (start < end && '._-'.includes(text[end - 1]!)) {
    end--;
  }

  while (start < end && shouldTrimEdgePunctuation(text[start]!)) {
    start++;
  }
  while (start < end && shouldTrimEdgePunctuation(text[end - 1]!)) {
    end--;
  }

  return { start, end, matchedText: text.slice(start, end) };
}

function finalizeSpan(original: string, start: number, end: number): OriginalSpan {
  return trimProfaneSpanEdges(original, start, end);
}

function fallbackSpan(
  original: string,
  variant: string,
  variantStart: number,
  variantEnd: number,
): OriginalSpan {
  const needle = variant.slice(variantStart, variantEnd);
  if (!needle) {
    return { start: 0, end: 0, matchedText: '' };
  }

  const lowerOriginal = original.toLowerCase();
  const lowerNeedle = needle.toLowerCase();
  const idx = lowerOriginal.indexOf(lowerNeedle);
  if (idx >= 0) {
    return finalizeSpan(original, idx, idx + needle.length);
  }

  // The variant word is not literally present in the original (e.g. it was
  // reconstructed from a fully masked token like "f******"). Returning a
  // guessed coordinate slice here only produces garbage spans, so drop it.
  return { start: 0, end: 0, matchedText: '' };
}

/**
 * Walk original and variant in parallel to locate the original span
 * corresponding to [variantStart, variantEnd) in the variant string.
 */
export function mapVariantSpanToOriginal(
  original: string,
  variant: string,
  variantStart: number,
  variantEnd: number,
): OriginalSpan {
  if (variantStart >= variantEnd) {
    return { start: 0, end: 0, matchedText: '' };
  }

  if (original === variant) {
    return finalizeSpan(original, variantStart, variantEnd);
  }

  let originalIndex = 0;
  let variantIndex = 0;
  let origStart = -1;
  let origEnd = -1;
  let lastAlignedNorm = '';
  let skippedSeparator = false;

  while (variantIndex < variant.length && originalIndex <= original.length) {
    if (variantIndex === variantStart && origStart === -1) {
      origStart = originalIndex;
      lastAlignedNorm = '';
      skippedSeparator = false;
    }
    if (variantIndex === variantEnd) {
      origEnd = originalIndex;
      break;
    }

    if (originalIndex >= original.length) {
      variantIndex++;
      continue;
    }

    const variantChar = variant[variantIndex]!;
    const originalChar = original[originalIndex]!;

    if (isCombiningMark(originalChar)) {
      originalIndex++;
      continue;
    }

    if (charsAlign(originalChar, variantChar)) {
      lastAlignedNorm = normalizeCharForAlign(originalChar) || originalChar.toLowerCase();
      skippedSeparator = false;
      variantIndex++;
      originalIndex++;
      continue;
    }

    if (isSkippableOriginalChar(originalChar)) {
      originalIndex++;
      skippedSeparator = true;
      continue;
    }

    // An extra original word character does not match the variant char.
    // Only tolerate it when it is a collapsed repeat ("fuuuck" -> "fuck")
    // that directly follows the previously aligned char with no separator
    // in between. Otherwise the variant char was removed/masked in the
    // original ("f******" -> "fuck") or positions drifted (NFKD punctuation
    // like "…" -> ".."): stop and let the caller fall back to a literal
    // lookup instead of consuming unrelated words.
    const originalNorm = normalizeCharForAlign(originalChar) || originalChar.toLowerCase();
    if (!skippedSeparator && lastAlignedNorm && originalNorm === lastAlignedNorm) {
      originalIndex++;
      continue;
    }

    break;
  }

  if (origEnd === -1 && variantIndex >= variantEnd) {
    origEnd = originalIndex;
  }

  if (origStart === -1 || origEnd === -1) {
    return fallbackSpan(original, variant, variantStart, variantEnd);
  }

  if (origStart >= origEnd) {
    return fallbackSpan(original, variant, variantStart, variantEnd);
  }

  const matchedText = original.slice(origStart, origEnd);
  if (!matchedText) {
    return fallbackSpan(original, variant, variantStart, variantEnd);
  }

  return finalizeSpan(original, origStart, origEnd);
}

export interface ProfaneSpan {
  word: string;
  start: number;
  end: number;
}

/** Keep longest original-text span when starts or ranges overlap. */
export function dedupeProfaneSpansByOverlap(
  spans: Array<[string, number, number]>,
): Array<[string, number, number]> {
  if (spans.length === 0) {
    return [];
  }

  const ordered = [...spans].sort(
    (left, right) => left[1] - right[1] || right[2] - right[1] - (left[2] - left[1]),
  );
  let kept: Array<[string, number, number]> = [];

  for (const candidate of ordered) {
    const [word, start, end] = candidate;
    const length = end - start;

    if (
      kept.some(
        ([, keptStart, keptEnd]) =>
          start >= keptStart && end <= keptEnd && keptEnd - keptStart > length,
      )
    ) {
      continue;
    }

    kept = kept.filter(
      ([, keptStart, keptEnd]) =>
        !(keptStart >= start && keptEnd <= end && length > keptEnd - keptStart),
    );
    kept.push([word, start, end]);
  }

  return kept;
}

/** True when `shorter` is a word-bounded substring of `longer` (avoids ass/classic). */
export function isNestedProfaneSpan(shorter: string, longer: string): boolean {
  if (longer.length <= shorter.length) {
    return false;
  }

  let searchFrom = 0;
  while (searchFrom <= longer.length - shorter.length) {
    const idx = longer.indexOf(shorter, searchFrom);
    if (idx === -1) {
      return false;
    }

    const beforeOk = idx === 0 || !isUnicodeWordChar(longer[idx - 1]!);
    const afterIdx = idx + shorter.length;
    const afterOk =
      afterIdx === longer.length || !isUnicodeWordChar(longer[afterIdx]!);

    if (beforeOk && afterOk) {
      return true;
    }

    searchFrom = idx + 1;
  }

  return false;
}
